libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp

```diff
@@ -261,59 +264,66 @@ struct message_receiver
+            auto f_send = [&] {
+                header hcopy = buf->header_;
+
+                // prepare new send buffer for forwarding message on
+                auto buf2 = get_msg_buffer(hcopy);
+                msg_info(rank, size, msg_type::send, buf2->header_, "send_R");
+                auto tx_snd2 = ex::just(buf2, message_size, MPI_UNSIGNED_CHAR,
+                                   next_rank(rank, size), tag, MPI_COMM_WORLD) |
+                    mpix::transform_mpi(MPI_Isend /*, mpix::stream_type::send_2*/) |
+                    ex::then([buf2 = buf2, rank = rank, size = size](/*int result*/) {
+                        counter--;
+                        msg_info(rank, size, msg_type::send, buf2->header_, "forwarded");
+                        release_msg_buffer(buf2);
+                    });
+
+                // launch the forwarding send for the current round
                 msr_deb<6>.debug(
-                    str<>("start_detached"), "rx_snd2", buf->header_.round, buf->header_.step);
-                ex::start_detached(std::move(rx_snd2));
+                    str<>("start_detached"), "tx_snd2", buf2->header_.round, buf2->header_.step);
+                ex::start_detached(std::move(tx_snd2));
+            };
+
+            auto f_recv = [&] {
+                // origin rank (step==0) receives one more than the other steps/ranks
+                if (buf->header_.step == 0 || (buf->header_.round < (num_rounds - 1)))
+                {
+                    msg_info(rank, size, msg_type::recv, buf->header_, "recv_R");
+                    // post a receive to handle it when it comes back on the next round
+                    // we can reuse the same buffer for the forwarding send and for the receive,
+                    // because the send _must_ complete before the receive
+                    // even if there is only one rank in the ring
+                    message_receiver reclambda(
+                        rank, orank, size, tag, num_rounds, message_size, buf);
+
+                    // the recursive lambda will handle it
+                    auto rx_snd2 = ex::just(buf, message_size, MPI_UNSIGNED_CHAR,
+                                       prev_rank(rank, size), tag, MPI_COMM_WORLD) |
+                        mpix::transform_mpi(MPI_Irecv /*, mpix::stream_type::receive_2*/) |
+                        ex::then(std::move(reclambda));
+                    // launch the receive for the msg on the next round
+                    msr_deb<6>.debug(
+                        str<>("start_detached"), "rx_snd2", buf->header_.round, buf->header_.step);
+                    ex::start_detached(std::move(rx_snd2));
+                }
+                else { release_msg_buffer(buf); }
+            };
+
+            if (recv_before_send)
+            {
+                f_recv();
+                f_send();
+            }
+            else
+            {
+                f_send();
+                f_recv();
             }
```

Comment:
much cleaner to read. thanks

Reply:
I considered adding an enum instead of the current "override default behaviour flag", something like `--mpi-order=send-before-recv|recv-before-send`. Would that be nicer, or is the current flag ok?

Reply:
Ideally we would not have a flag at all, just change to the version that works. Leave it as you've done it. In general, pre-posting recvs is a good thing, which is why I did it that way, but the fact that is locks up makes it bad. So merge this as is and in the next round of concolidation/cleaning I'll reassess it.